File: matlab/tools/bss/spectral_component_analysis_dft.m
function [y,W,A] = spectral_component_analysis_dft(x,fl,fu)
% spectral_component_analysis_dft - Spectral component analysis (SCA) using
% a Fourier domain approach. Extracts and ranks linear mixtures of
% multichannel data with maximal energy in a given frequency band.
%
% Usage:
%   [y,W,A] = spectral_component_analysis_dft(x,fl,fu)
%
% Inputs:
%   x: Input data array (channels x samples).
%   fl: lower cutoff frequency normalized by sampling frequency
%   fu: upper cutoff frequency normalized by sampling frequency
%
% Outputs:
%   y: Extracted spectral components ranked by their energy in the
%       frequency band of interest
%   W: Extraction (separation) matrix
%   A: Mixing matrix
%
% Reference: - Sameni, R., Jutten, C., & Shamsollahi, M. B. (2009). A
%   deflation procedure for subspace decomposition. IEEE Transactions on
%   Signal Processing, 58(4), 2363-2374.
%
% Revision History:
%   2009: First release
%   2023: Renamed from deprecated version SCA. corrected a frequency
%       binning bug
%
% Reza Sameni, 2009-2023 The Open-Source Electrophysiological Toolbox
% https://github.com/alphanumericslab/OSET

L2 = size(x,2);

freq_indexes = max(round(L2*fl), 1) : min(round(L2*fu), L2); % indexes of the frequency range of interest (half band only)

X = fft(x, L2, 2); % Fourier transform of the input data
Bf = X(:, freq_indexes) * X(:, freq_indexes)'; % Compute the spectral covariance matrix
B = 2 * real(Bf); % Multiply by 2 to compensate for counting only half the signal power in the frequency range of interest
B = (B + B') / 2; % Make the matrix symmetric

C = x * x'; % Compute the covariance matrix
C = (C + C') / 2; % Make the matrix symmetric

[V, D] = eig(B, C, 'chol'); % Perform eigenvalue decomposition using Cholesky decomposition

[~, II] = sort(diag(D), 'descend'); % Sort eigenvalues in descending order

W = V(:, II)'; % Extraction matrix
A = pinv(W); % Mixing matrix (pseudo-inverse of the extraction matrix)

y = real(W * x); % Extract spectral components by applying the extraction matrix to the input data
end

File: python/src/oset/ecg/bss/spectral_component_analysis_dft.py
import numpy as np
from scipy.linalg import eigh
import argparse

def spectral_component_analysis_dft(x, fl, fu):
    """
    Spectral component analysis (SCA) using a Fourier domain approach.
    Extracts and ranks linear mixtures of multichannel data with maximal energy
    in a given frequency band.

    Parameters:
    x (ndarray): Input data array (channels x samples).
    fl (float): Lower cutoff frequency normalized by sampling frequency.
    fu (float): Upper cutoff frequency normalized by sampling frequency.

    Returns:
    tuple: A tuple containing:
        y (ndarray): Extracted spectral components ranked by their energy in the frequency band of interest.
        W (ndarray): Extraction (separation) matrix.
        A (ndarray): Mixing matrix.

    Reference:
        Sameni, R., Jutten, C., & Shamsollahi, M. B. (2009). A deflation procedure for subspace decomposition. 
        IEEE Transactions on Signal Processing, 58(4), 2363-2374.
    
    Revision History:
        July 2024: Translated to Python from Matlab (spectral_component_analysis_dft.m)

    Muhammad Ubadah Tanveer, 2024
    The Open-Source Electrophysiological Toolbox
    https://github.com/alphanumericslab/OSET
    """

    L2 = x.shape[1]

    freq_indexes = np.arange(max(round(L2 * fl), 1), min(round(L2 * fu), L2) + 1)

    X = np.fft.fft(x, L2, axis=1)
    Bf = X[:, freq_indexes] @ X[:, freq_indexes].conj().T
    B = 2 * np.real(Bf)
    B = (B + B.T) / 2

    C = x @ x.T
    C = (C + C.T) / 2

    D, V = eigh(B, C)
    II = np.argsort(D)[::-1]

    W = V[:, II].T
    A = np.linalg.pinv(W)

    y = np.real(W @ x)

    return y, W, A


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="""
    Spectral component analysis (SCA) using a Fourier domain approach.
    Extracts and ranks linear mixtures of multichannel data with maximal energy
    in a given frequency band.

    Parameters:
    x (ndarray): Input data array (channels x samples).
    fl (float): Lower cutoff frequency normalized by sampling frequency.
    fu (float): Upper cutoff frequency normalized by sampling frequency.

    Returns:
    tuple: A tuple containing:
        y (ndarray): Extracted spectral components ranked by their energy in the frequency band of interest.
        W (ndarray): Extraction (separation) matrix.
        A (ndarray): Mixing matrix.

    Revision History:
        June 2024: Translated to Python from Matlab (spectral_component_analysis_dft.m)
    """,
        formatter_class=argparse.RawTextHelpFormatter,
    )
    args = parser.parse_args()